net_util/src/queue_pair.rs

```diff
@@ -288,6 +292,44 @@ impl RxVirtio {
     }
 }
 
+#[derive(Default, Clone)]
+struct IovecBuffer(Vec<libc::iovec>);
+
+unsafe impl Send for IovecBuffer {}
+unsafe impl Sync for IovecBuffer {}
+
+impl IovecBuffer {
+    const fn new() -> Self {
+        IovecBuffer(Vec::new())
+    }
+
+    fn borrow(&mut self) -> IovecBufferBorrowed<'_> {
+        IovecBufferBorrowed(&mut self.0)
+    }
+}
+
+struct IovecBufferBorrowed<'a>(&'a mut Vec<libc::iovec>);
+
+impl<'a> std::ops::Deref for IovecBufferBorrowed<'a> {
+    type Target = Vec<libc::iovec>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<'a> std::ops::DerefMut for IovecBufferBorrowed<'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
+}
+
+impl Drop for IovecBufferBorrowed<'_> {
+    fn drop(&mut self) {
+        self.0.clear();
```

Comment:
My Rust knowledge on this part is a bit rusty (pun intended), but why `clear` here?

Do you even need to implement `Drop`? When IovecBuffer is dropped, its only member will also get dropped, right?

Reply:
Here `IovecBufferBorrowed` is a zero-cost wrapper for a mutable borrowed Vec. It is introduced to impl drop manually(instead of just returning `&mut Vec<_>`).

Clear the Vec when drop can make sure there's no old value inside for the next user. With the Rust borrow checker, we can ensure that the buffer is not used by anyone else between a single {write, syscall, clear}, and that no old value remains when it is used next time.

This make it works just like creating a new Vec.
- `IovecBuffer::new` -> `Vec::new`
- `IovecBufferBorrowed::some_fn` -> `Vec::some_fn` because of Deref/DerefMut
- `IovecBufferBorrowed::drop` -> `Vec::drop`

Reply:
I see. I misread the code. I thought you were implementing `Drop` for `IovecBuffer`.